Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py

```diff
@@ -37,48 +68,77 @@ def validateInputs(self):
-                             doc='The mask from this workspace will be applied before reduction')
-
-        self.copyProperties('ConvertSpectrumAxis', ['Target', 'EFixed'])
-
-        self.copyProperties('ResampleX', ['XMin', 'XMax', 'NumberBins', 'LogBinning'])
-
-        self.declareProperty('NormaliseBy', 'Monitor', StringListValidator(['None', 'Time', 'Monitor']), "Normalise to monitor or time. ")
-
-        self.declareProperty('MaskAngle', Property.EMPTY_DBL,
-                             "Phi angle above which will be masked. See :ref:`MaskAngle <algm-MaskAngle>` for details.")
-
-        self.declareProperty(MatrixWorkspaceProperty("OutputWorkspace", "",
-                                                     direction=Direction.Output),
-                             doc="Output Workspace")
+        self.declareProperty(
+            StringArrayProperty(
+                "InputWorkspace", direction=Direction.Input, validator=ADSValidator(),
+            ),
+            doc="The main input workspace[s].",
+        )
+
+        self.declareProperty(
+            MatrixWorkspaceProperty(
+                "BackgroundWorkspace",
+                "",
+                optional=PropertyMode.Optional,
+                direction=Direction.Input,
+            ),
+            doc="The background workspace to be subtracted.",
+        )
+
+        self.declareProperty(
+            MatrixWorkspaceProperty(
+                "CalibrationWorkspace",
+                "",
+                optional=PropertyMode.Optional,
+                direction=Direction.Input,
+            ),
+            doc="The calibration (vandiaum) workspace.",
+        )
+
+        self.declareProperty(
+            "BackgroundScale",
+            1.0,
+            validator=FloatBoundedValidator(0.0),
+            doc="The background will be scaled by this number before being subtracted.",
+        )
+
+        self.declareProperty(
+            MaskWorkspaceProperty(
+                "MaskWorkspace",
+                "",
+                optional=PropertyMode.Optional,
+                direction=Direction.Input,
+            ),
+            doc="The mask from this workspace will be applied before reduction",
+        )
+
+        self.copyProperties("ConvertSpectrumAxis", ["Target", "EFixed"])
+
+        self.copyProperties("ResampleX", ["XMin", "XMax", "NumberBins", "LogBinning"])
+
+        self.declareProperty(
+            "NormaliseBy",
+            "Monitor",
+            StringListValidator(["None", "Time", "Monitor"]),
+            "Normalise to monitor or time. ",
+        )
+
+        self.declareProperty(
+            "MaskAngle",
+            Property.EMPTY_DBL,
+            "Phi angle above which will be masked. See :ref:`MaskAngle <algm-MaskAngle>` for details.",
+        )
+
+        self.declareProperty(
+            MatrixWorkspaceProperty("OutputWorkspace", "", direction=Direction.Output),
+            doc="Output Workspace",
+        )
 
     def PyExec(self):
```

Comment:
Overall, method PyExec is very long.  It is better to refactor it into several methods.

Reply:
`PyExec` is now refactored and more concise than the previous version.